nav2-usability/scenario-code/nested-routing/router.dart

```diff
@@ -0,0 +1,353 @@
+
+    if (path is NewBooksRoutePath) {
+      location = '/books/new';
+    } else if (path is AllBooksRoutePath) {
+      location = '/books/all';
+    } else if (path is SettingsRoutePath) {
+      location = '/settings';
+    }
+
+    return RouteInformation(location: location);
+  }
+}
+
+class BookRouterDelegate extends RouterDelegate<AppRoutePath>
+    with ChangeNotifier, PopNavigatorRouterDelegateMixin<AppRoutePath> {
+  @override
+  final GlobalKey<NavigatorState> navigatorKey;
+  final AppState _appState = AppState();
+  late final OverlayEntry _overlayEntry =
+  OverlayEntry(builder: _overlayEntryBuilder);
+
+  BookRouterDelegate() : navigatorKey = GlobalKey<NavigatorState>() {
+    _appState.addListener(handleAppStateChanged);
+  }
+
+  void handleAppStateChanged() {
+    notifyListeners();
+    _overlayEntry.markNeedsBuild();
+  }
+
+  @override
+  Future<void> setNewRoutePath(AppRoutePath path) async {
+    if (path is NewBooksRoutePath) {
+      _appState.page = AppPage.newBooks;
+    } else if (path is AllBooksRoutePath) {
+      _appState.page = AppPage.allBooks;
+    } else if (path is SettingsRoutePath) {
+      _appState.page = AppPage.settings;
+    }
+  }
+
+  @override
+  void dispose() {
+    _appState.dispose();
+    super.dispose();
+  }
+
+  @override
+  AppRoutePath get currentConfiguration {
+    if (_appState.page == AppPage.newBooks) {
+      return NewBooksRoutePath();
+    } else if (_appState.page == AppPage.allBooks) {
+      return AllBooksRoutePath();
+    } else if (_appState.page == AppPage.settings) {
+      return SettingsRoutePath();
+    }
+    return AppRoutePath();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Overlay(
+      initialEntries: [_overlayEntry],
+    );
+  }
+
+  Widget _overlayEntryBuilder(BuildContext context) {
+    return Scaffold(
+      body: Navigator(
+        key: navigatorKey,
+        pages: [
+          if (_appState.page == AppPage.settings)
+            FadeTransitionPage(
+              key: ValueKey('SettingsScreen'),
+              child: SettingsScreen(),
+            ),
+          if (_appState.page == AppPage.newBooks ||
+              _appState.page == AppPage.allBooks)
+            FadeTransitionPage(
+              key: ValueKey('BooksScreen'),
+              child: BooksScreen(
+                onTabSelected: (int idx) {
+                  if (idx == 0) {
+                    _appState.page = AppPage.newBooks;
+                  } else {
+                    _appState.page = AppPage.allBooks;
+                  }
+                },
+                appPage: _appState.page,
+              ),
+            ),
+        ],
+        onPopPage: (route, result) {
+          return route.didPop(result);
+        },
+      ),
+      bottomNavigationBar: BottomNavigationBar(
+        currentIndex: _appState.page == AppPage.settings ? 1 : 0,
+        onTap: (idx) {
+          if (idx == 0) {
+            _appState.page = AppPage.newBooks;
```

Comment:
One way to solve the issue that https://github.com/flutter/uxr/pull/34#issuecomment-823623869 mentioned is to  keep track of two boolean value in the _appState
```
bool inBook;
bool filterByNew;
```
instead of storing the page enum.

This way you just need to set the inBook to true/false in this callback while preserving the filter value